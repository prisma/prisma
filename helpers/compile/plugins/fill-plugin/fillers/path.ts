export function resolve(...args: any[]) {
  return args.join('/')
}

export function join(...args: any[]) {
  return args.join('/')
}

export const sep = '/'

export const posix = {
  sep,
}

/**
 * A poor man's shim for the "path" module
 */
const path = {
  resolve,
  posix,
  join,
  sep,
}

export default path
