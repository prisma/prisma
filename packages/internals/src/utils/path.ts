import path from 'path'

/**
 * Normalize `filePath` to use forward slashes as a separator. `filePath` is
 * treated as a path specific to the current platform, so backslashes will only
 * be replaced with forward slashes on Windows. On other platforms, where a
 * backslash is a valid filename character, it will be treated as such and will
 * not be replaced.
 */
export function pathToPosix(filePath: string): string {
  if (path.sep === path.posix.sep) {
    return filePath
  }
  return filePath.split(path.sep).join(path.posix.sep)
}

/**
 * Returns the longest common path ancestor of the two paths (which may also be equal to one or both of them).
 * If they don't share a common ancestor (which may be the case on Windows if both paths are on different disks),
 * returns `undefined`.
 */
export function longestCommonPathPrefix(pathA: string, pathB: string): string | undefined {
  if (!path.isAbsolute(pathA) || !path.isAbsolute(pathB)) {
    throw new Error('longestCommonPathPrefix expects absolute paths')
  }

  if (process.platform === 'win32' && (pathA.startsWith('\\\\') || pathB.startsWith('\\\\'))) {
    // Make both paths namespaced if at least one of them is.
    pathA = path.toNamespacedPath(pathA)
    pathB = path.toNamespacedPath(pathB)
  }

  const commonPrefix = longestCommonPrefix(pathA.split(path.sep), pathB.split(path.sep)).join(path.sep)

  if (commonPrefix === '') {
    return process.platform === 'win32' ? undefined : '/'
  }

  if (process.platform === 'win32' && ['\\', '\\\\?', '\\\\.'].includes(commonPrefix)) {
    return undefined
  }

  if (process.platform === 'win32' && commonPrefix.endsWith(':')) {
    // Disk specifier without a backslash at the end is not an absolute path on windows,
    // it refers to current working directory on that disk.
    return commonPrefix + '\\'
  }

  return commonPrefix
}

function longestCommonPrefix<T>(sequenceA: T[], sequenceB: T[]): T[] {
  const maxLen = Math.min(sequenceA.length, sequenceB.length)
  let sliceLen = 0

  while (sliceLen <= maxLen && sequenceA[sliceLen] === sequenceB[sliceLen]) {
    sliceLen++
  }

  return sequenceA.slice(0, sliceLen)
}
