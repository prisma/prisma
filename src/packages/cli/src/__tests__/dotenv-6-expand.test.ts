import path from 'path'
import assert from 'assert'
import stripAnsi from 'strip-ansi'

it('should read expanded env vars', async () => {
  process.argv.push('--version')

  const oldConsoleLog = console.error
  const logs: string[] = []
  console.error = (...args) => {
    logs.push(...args)
  }

  const cwd = process.cwd()
  process.chdir(path.join(__dirname, './fixtures/dotenv-6-expand'))
  process.argv.push('--schema=./expand/schema.prisma')
  await import('../bin')
  console.error = oldConsoleLog

  assert.equal(
    stripAnsi(logs.join()),
    'Environment variables loaded from provided --schema directory',
  )

  assert.equal(
    process.env.DOTENV_PRISMA_EXPAND_DATABASE_URL_WITH_SCHEMA,
    'postgres://user:password@server.host:5432/database?ssl=1&schema=schema1234',
    'process.env.DOTENV_PRISMA_EXPAND_DATABASE_URL_WITH_SCHEMA',
  )

  process.chdir(cwd)
})
