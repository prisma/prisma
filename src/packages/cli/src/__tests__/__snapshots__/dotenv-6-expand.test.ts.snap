// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should read expanded env vars 1`] = `
Array [
  Array [
    Environment variables loaded from expand/.env,
  ],
]
`;
